File: src/android/KartePlugin.java
package io.karte.cordova;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.apache.cordova.PluginResult.Status;
import org.jetbrains.annotations.NotNull;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import io.karte.android.KarteApp;
import io.karte.android.core.library.Library;
import io.karte.android.core.usersync.UserSync;
import io.karte.android.inappmessaging.InAppMessaging;
import io.karte.android.notifications.Notifications;
import io.karte.android.tracking.Tracker;
import io.karte.android.variables.FetchCompletion;
import io.karte.android.variables.Variable;

public final class KartePlugin extends CordovaPlugin implements Library {
    private KarteVariables variables = new KarteVariables();

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) {
        try {
            if ("visitorId".equals(action)) {
                return this.visitorId(callbackContext);
            } else if ("isOptOut".equals(action)) {
                return this.isOptOut(callbackContext);
            } else if ("optIn".equals(action)) {
                return this.optIn(callbackContext);
            } else if ("optOut".equals(action)) {
                return this.optOut(callbackContext);
            } else if ("renewVisitorId".equals(action)) {
                return this.renewVisitorId(callbackContext);
            } else if ("appendingQueryParameter".equals(action)) {
                return this.appendingQueryParameter(callbackContext, args.getString(0));
            } else if ("track".equals(action)) {
                return this.track(callbackContext, args.getString(0), args.optJSONObject(1));
            } else if ("identify".equals(action)) {
                return this.identify(callbackContext, args.optJSONObject(0));
            } else if ("view".equals(action)) {
                return this.view(callbackContext, args.getString(0), args.optString(1), args.optJSONObject(2));
            } else if ("isPresenting".equals(action)) {
                return this.isPresenting(callbackContext);
            } else if ("dismiss".equals(action)) {
                return this.dismiss(callbackContext);
            } else if ("suppress".equals(action)) {
                return this.suppress(callbackContext);
            } else if ("unsuppress".equals(action)) {
                return this.unsuppress(callbackContext);
            } else if ("registerFCMToken".equals(action)) {
                return this.registerFCMToken(callbackContext, args.getString(0));
            } else if ("fetch".equals(action)) {
                return this.fetch(callbackContext);
            } else if ("variable".equals(action)) {
                return  this.variable(callbackContext, args.getString(0));
            } else if ("trackOpen".equals(action)) {
                return this.trackOpen(callbackContext, args.getJSONArray(0), args.optJSONObject(1));
            } else if ("trackClick".equals(action)) {
                return this.trackClick(callbackContext, args.getJSONArray(0), args.optJSONObject(1));
            } else if ("string".equals(action)) {
                return this.getString(callbackContext, args.getString(0), args.optString(1));
            } else if ("integer".equals(action)) {
                return this.getInteger(callbackContext, args.getString(0), args.getLong(1));
            } else if ("double".equals(action)) {
                return this.getDouble(callbackContext, args.getString(0), args.getDouble(1));
            } else if ("bool".equals(action)) {
                return this.getBoolean(callbackContext, args.getString(0), args.getBoolean(1));
            } else if ("array".equals(action)) {
                return this.getArray(callbackContext, args.getString(0), args.optJSONArray(1));
            } else if ("object".equals(action)) {
                return this.getObject(callbackContext, args.getString(0), args.optJSONObject(1));
            }
        } catch (Exception e) {
            handleException(e, callbackContext);
        }
        return false;
    }

    private boolean visitorId(final CallbackContext callbackContext) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                callbackContext.success(KarteApp.getVisitorId());
            }
        });
        return true;
    }

    private boolean isOptOut(final CallbackContext callbackContext) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                PluginResult pluginResult = new PluginResult(Status.OK, KarteApp.isOptOut());
                callbackContext.sendPluginResult(pluginResult);
            }
        });
        return true;
    }

    private boolean optIn(final CallbackContext callbackContext) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                KarteApp.optIn();
                callbackContext.success();
            }
        });
        return true;
    }

    private boolean optOut(final CallbackContext callbackContext) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                KarteApp.optOut();
                callbackContext.success();
            }
        });
        return true;
    }

    private boolean renewVisitorId(final CallbackContext callbackContext) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                KarteApp.renewVisitorId();
                callbackContext.success();
            }
        });
        return true;
    }

    private boolean appendingQueryParameter(final CallbackContext callbackContext, final String url) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                callbackContext.success(UserSync.appendUserSyncQueryParameter(url));
            }
        });
        return true;
    }

    private boolean track(final CallbackContext callbackContext, final String name, final JSONObject values) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Tracker.track(name, values);
                callbackContext.success();
            }
        });
        return true;
    }

    private boolean identify(final CallbackContext callbackContext, final JSONObject values) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Tracker.identify(values);
                callbackContext.success();
            }
        });
        return true;
    }

    private boolean view(final CallbackContext callbackContext, final String viewName, final String title, final JSONObject values) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Tracker.view(viewName, title, values);
                callbackContext.success();
            }
        });
        return true;
    }

    private boolean isPresenting(final CallbackContext callbackContext) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                PluginResult pluginResult = new PluginResult(Status.OK, InAppMessaging.isPresenting());
                callbackContext.sendPluginResult(pluginResult);
            }
        });
        return true;
    }

    private boolean dismiss(final CallbackContext callbackContext) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                InAppMessaging.dismiss();
                callbackContext.success();
            }
        });
        return true;
    }

    private boolean suppress(final CallbackContext callbackContext) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                InAppMessaging.suppress();
                callbackContext.success();
            }
        });
        return true;
    }

    private boolean unsuppress(final CallbackContext callbackContext) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                InAppMessaging.unsuppress();
                callbackContext.success();
            }
        });
        return true;
    }

    private boolean registerFCMToken(final CallbackContext callbackContext, final String fcmToken) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Notifications.registerFCMToken(fcmToken);
                callbackContext.success();
            }
        });
        return true;
    }

    private boolean fetch(final CallbackContext callbackContext) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                KartePlugin.this.variables.fetch(new FetchCompletion() {
                    @Override
                    public void onComplete(boolean success) {
                        if (success) {
                            callbackContext.success();
                        } else {
                            callbackContext.error("Failed to fetch.");
                        }
                    }
                });
            }
        });
        return true;
    }

    private boolean variable(final CallbackContext callbackContext, final String key) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Variable variable = variables.variable(key);
                callbackContext.success(variable.getName());
            }
        });
        return true;
    }

    private boolean trackOpen(final CallbackContext callbackContext, final JSONArray keys, final JSONObject values) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    KartePlugin.this.variables.trackOpen(keys, values);
                    callbackContext.success();
                } catch (JSONException e) {
                    handleException(e, callbackContext);
                }
            }
        });
        return true;
    }

    private boolean trackClick(final CallbackContext callbackContext, final JSONArray keys, final JSONObject values) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    KartePlugin.this.variables.trackClick(keys, values);
                    callbackContext.success();
                } catch (JSONException e) {
                    handleException(e, callbackContext);
                }
            }
        });
        return true;
    }

    private boolean getString(final CallbackContext callbackContext, final String key, final String defaultValue) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                String value = KartePlugin.this.variables.getString(key, defaultValue);
                callbackContext.success(value);
            }
        });
        return true;
    }

    private boolean getInteger(final CallbackContext callbackContext, final String key, final Long defaultValue) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                int value = (int)(long)KartePlugin.this.variables.getLong(key, defaultValue);
                callbackContext.success(value);
            }
        });
        return true;
    }

    private boolean getDouble(final CallbackContext callbackContext, final String key, final Double defaultValue) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                float value  =(float)(double)KartePlugin.this.variables.getDouble(key, defaultValue);
                PluginResult pluginResult = new PluginResult(Status.OK, value);
                callbackContext.sendPluginResult(pluginResult);
            }
        });
        return true;
    }

    private boolean getBoolean(final CallbackContext callbackContext, final String key, final Boolean defaultValue) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Boolean value = KartePlugin.this.variables.getBoolean(key, defaultValue);
                PluginResult pluginResult = new PluginResult(Status.OK, value);
                callbackContext.sendPluginResult(pluginResult);
            }
        });
        return true;
    }

    private boolean getArray(final CallbackContext callbackContext, final String key, final JSONArray defaultValue) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                JSONArray value = KartePlugin.this.variables.getJSONArray(key, defaultValue);
                callbackContext.success(value);
            }
        });
        return true;
    }

    private boolean getObject(final CallbackContext callbackContext, final String key, final JSONObject defaultValue) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                JSONObject value = KartePlugin.this.variables.getJSONObject(key, defaultValue);
                callbackContext.success(value);
            }
        });
        return true;
    }

    private static void handleException(Exception e, CallbackContext callbackContext) {
        String message = e.toString();
        callbackContext.error(message);
    }

    // implements Library
    @Override
    public boolean isPublic() {
        return true;
    }

    @NotNull
    @Override
    public String getName() {
        return "cordova";
    }

    @NotNull
    @Override
    public String getVersion() {
        return "0.0.1";
    }

    @Override
    public void configure(@NotNull KarteApp karteApp) {
    }

    @Override
    public void unconfigure(@NotNull KarteApp karteApp) {
    }
}

File: src/ios/KartePlugin.swift
import KarteCore
import KarteInAppMessaging
import KarteRemoteNotification

@objc(KartePlugin)
class KartePlugin: CDVPlugin {
    private lazy var variables: KarteVariables = {
        KarteVariables()
    }()
    
    // MARK: KarteCore
    @objc(visitorId:)
    func visitorId(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            let result = CDVPluginResult(status: .ok, messageAs: KarteApp.visitorId)
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
    @objc(isOptOut:)
    func isOptOut(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            let result = CDVPluginResult(status: .ok, messageAs: KarteApp.isOptOut)
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
    @objc(optIn:)
    func optIn(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            KarteApp.optIn()
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }
    
    @objc(optOut:)
    func optOut(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            KarteApp.optOut()
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }
    
    @objc(renewVisitorId:)
    func renewVisitorId(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            KarteApp.renewVisitorId()
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }
    
    @objc(registerFCMToken:)
    func registerFCMToken(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            let fcmToken = command.argument(at: 0) as? String
            KarteApp.registerFCMToken(fcmToken)
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }
    
    override func handleOpenURL(_ notification: Notification!) {
        guard let url = notification.object as? URL else {
            return
        }
        KarteApp.application(UIApplication.shared, open: url)
    }
    
    @objc(appendingQueryParameter:)
    func appendingQueryParameter(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let url = command.argument(at: 0) as? String else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            
            let result = CDVPluginResult(status: .ok, messageAs: UserSync.appendingQueryParameter(url))
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }

    @objc(getUserSyncScript:)
    func getUserSyncScript(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            let result = CDVPluginResult(status: .ok, messageAs: UserSync.getUserSyncScript())
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
    // MARK: KarteCore (Tracking)
    @objc(track:)
    func track(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let name = command.argument(at: 0) as? String else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            let values = command.argument(at: 1) as? [String: Any] ?? [:]
            
            Tracker.track(name, values: JSONConvertibleConverter.convert(values))
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }
    
    @objc(identify:)
    func identify(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            let values = command.argument(at: 0) as? [String: Any] ?? [:]
            
            Tracker.identify(JSONConvertibleConverter.convert(values))
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }

    @objc(identifyWithUserId:)
    func identifyWithUserId(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let userId = command.argument(at: 0) as? String else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            let values = command.argument(at: 1) as? [String: Any] ?? [:]

            Tracker.identify(userId, JSONConvertibleConverter.convert(values))
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }

    @objc(attribute:)
    func attribute(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            let values = command.argument(at: 0) as? [String: Any] ?? [:]

            Tracker.attribute(JSONConvertibleConverter.convert(values))
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }
    
    @objc(view:)
    func view(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let viewName = command.argument(at: 0) as? String else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            
            let title = command.argument(at: 1) as? String
            let values = command.argument(at: 2) as? [String: Any] ?? [:]
            
            Tracker.view(viewName, title: title, values: JSONConvertibleConverter.convert(values))
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }
    
    // MARK: KarteInAppMessaging
    @objc(isPresenting:)
    func isPresenting(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            let result = CDVPluginResult(status: .ok, messageAs: InAppMessaging.shared.isPresenting)
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
    @objc(dismiss:)
    func dismiss(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            InAppMessaging.shared.dismiss()
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }
    
    @objc(suppress:)
    func suppress(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            InAppMessaging.shared.suppress()
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }
    
    @objc(unsuppress:)
    func unsuppress(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            InAppMessaging.shared.unsuppress()
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }
    
    // MARK: KarteVariables
    @objc(fetch:)
    func fetch(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            self.variables.fetch { [weak self] (isSuccessful) in
                let result = isSuccessful ? CDVPluginResult(status: .ok) : CDVPluginResult(status: .error, messageAs: "Failed to fetch.")
                self?.commandDelegate.send(result, callbackId: command.callbackId)
            }
        }
    }
    
    @objc(variable:)
    func variable(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let key = command.argument(at: 0) as? String else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            
            let variable = self.variables.variable(forKey: key)
            let result = CDVPluginResult(status: .ok, messageAs: variable.name)
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
    @objc(trackOpen:)
    func trackOpen(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let keys = command.argument(at: 0) as? [String] else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            let values = command.argument(at: 1) as? [String: Any] ?? [:]
            self.variables.trackOpen(forKeys: keys, values: values)
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }
    
    @objc(trackClick:)
    func trackClick(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let keys = command.argument(at: 0) as? [String] else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            let values = command.argument(at: 1) as? [String: Any] ?? [:]
            self.variables.trackClick(forKeys: keys, values: values)
            self.commandDelegate.send(.noResult, callbackId: command.callbackId)
        }
    }
    
    // MARK: KarteVariable
    @objc(string:)
    func string(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let key = command.argument(at: 0) as? String else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            
            let string: String?
            if let value = command.argument(at: 1) as? String {
                string = self.variables.string(forKey: key, default: value)
            } else {
                string = self.variables.string(forKey: key)
            }
            
            let result = CDVPluginResult(status: .ok, messageAs: string)
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
    @objc(integer:)
    func integer(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let key = command.argument(at: 0) as? String else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            guard let value = command.argument(at: 1) as? Int else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            
            let result = CDVPluginResult(status: .ok, messageAs: self.variables.integer(forKey: key, default: value))
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
    @objc(double:)
    func double(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let key = command.argument(at: 0) as? String else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            guard let value = command.argument(at: 1) as? Double else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            
            let result = CDVPluginResult(status: .ok, messageAs: self.variables.double(forKey: key, default: value))
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
    @objc(bool:)
    func bool(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let key = command.argument(at: 0) as? String else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            guard let value = command.argument(at: 1) as? Bool else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            
            let result = CDVPluginResult(status: .ok, messageAs: self.variables.bool(forKey: key, default: value))
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
    @objc(array:)
    func array(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let key = command.argument(at: 0) as? String else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            
            let array: [Any]?
            if let value = command.argument(at: 1) as? [Any] {
                array = self.variables.array(forKey: key, default: value)
            } else {
                array = self.variables.array(forKey: key)
            }
            
            let result = CDVPluginResult(status: .ok, messageAs: array)
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
    @objc(object:)
    func object(command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let key = command.argument(at: 0) as? String else {
                self.commandDelegate.send(.errorMissingArgument, callbackId: command.callbackId)
                return
            }
            
            let object: [String: Any]?
            if let value = command.argument(at: 1) as? [String: Any] {
                object = self.variables.object(forKey: key, default: value)
            } else {
                object = self.variables.object(forKey: key)
            }
            
            let result = CDVPluginResult(status: .ok, messageAs: object)
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
}

extension KartePlugin: Library {
    static var name: String {
        "cordova"
    }
    
    static var version: String {
        "0.0.1"
    }
    
    static var isPublic: Bool {
        true
    }
    
    static func configure(app: KarteApp) {
    }
    
    static func unconfigure(app: KarteApp) {
    }
}

extension KartePluginLoader {
    open override class func handleLoad() {
        KarteApp.register(library: KartePlugin.self)
    }
}

private extension CDVPluginResult {
    
    static var noResult: CDVPluginResult {
        CDVPluginResult()
    }
    
    static var errorMissingArgument: CDVPluginResult {
        CDVPluginResult(status: .error, messageAs: "Argument is missing.")
    }
}
